CONTRACTS: Dynamic frames for loop contracts

qinheping opened the pull request:
This PR adds support for loop contracts with dynamic frames.

The first two commits are from #7671  #7672, please review separately.

Fixes #7196 
Fixes #7517 

We modify the `dfcc_instrument` class to use `dfcc_instrument_loopt` and apply loop contracts transformation at the same time as dynamic frame instrumentation, when the `--apply-loop-contract-flag` is set.

We also port the tests from the previous loop contracts implementation to the new one, and add some tests that combine loop contracts and recursive functions.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp:
```diff
@@ -25,6 +25,7 @@ Author: Remi Delmas, delmarsd@amazon.com
 #include <ansi-c/c_object_factory_parameters.h>
 #include <goto-instrument/contracts/cfg_info.h>
 #include <goto-instrument/contracts/contracts.h>
+#include <goto-instrument/contracts/instrument_spec_assigns.h>
```
this should not be used anymore, this is the old static instrumentation

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp:
```diff
@@ -338,6 +340,764 @@ void dfcc_instrumentt::instrument_function(
+  const goto_programt::instructiont::targett target)
+{
+  const auto &pragma = get_loop_id_pragma(target);
+  return pragma != empty_pragma;
+}
+
+void dfcc_instrumentt::tag_loop_instruction(
+  goto_programt::instructiont::targett &target,
+  const irep_idt &pragma,
+  const size_t loop_id)
+{
+  target->source_location_nonconst().add_pragma(
+    id2string(pragma) + ":" + std::to_string(loop_id));
+}
+
+irep_idt
+dfcc_instrumentt::get_loop_id_pragma(const goto_programt::targett &target)
+{
+  for(const auto &pragma : target->source_location().get_pragmas())
+  {
+    if(is_loop_id_pragma(pragma.first))
+      return pragma.first;
+  }
+  return empty_pragma;
+}
+
+void dfcc_instrumentt::remove_loop_id_pragma(goto_programt::targett &target)
+{
+  for(const auto &pragma : target->source_location().get_pragmas())
+  {
+    if(is_loop_id_pragma(pragma.first))
+    {
+      target->source_location_nonconst()
+        .get_named_sub()
+        .add(ID_pragma)
+        ->second.remove(pragma.first);
+      return;
+    }
+  }
+}
```
Could we avoid looping over all pragmas and having to parse the string each time ?

How about using something like this:

```
const irep_idt ID_loop_id = "loop-id";
const irep_idt ID_loop_instr_type = "loop-instr-type";
const irep_idt ID_loop_instr_body = "body";
const irep_idt ID_loop_instr_head = "head";
const irep_idt ID_loop_instr_latch = "latch";
const irep_idt ID_loop_instr_exiting = "exiting";

void tag_loop_instr(targett &target, const irep_idt &instr_type, const std::size_t loop_id)
{
    irept &irep = target->source_location_nonconst();
    irep.set(ID_loop_instr_type, instr_type);
    irep.set(ID_loop_id, loop_id);
}

irep_idt &get_loop_instr(targett &target)
{
    irept &irep = target->source_location();
    return irep.get(ID_loop_instr_type);
}

std::size_t get_loop_id(targett &target)
{
    PRECONDITION(!get_loop_instr_type(target).empty());
    irept &irep = target->source_location();
    return irep.get_size_t(ID_loop_id);
}

bool is_loop_instr(targett &target, std::size_t loop_id) {
    if (get_loop_instr_type(target).empty()) return false;
    return get_loop_id(target) == loop_id;
}
```

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp:
```diff
@@ -338,6 +340,764 @@ void dfcc_instrumentt::instrument_function(
+  }
+}
+
+void dfcc_instrumentt::compute_loop_nesting_graph(
```
Could we move this method and the loop graph node type declaration to a `dfcc_loop_utilst` class in a separate file ?

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.h:
```diff
@@ -34,6 +41,48 @@ class cfg_infot;
 class dfcc_libraryt;
 class dfcc_utilst;
 
+/// A local pragma used to keep track instructions in loops.
+const char CONTRACT_PRAGMA_loop_id[] = "contracts:loop-id";
+const char CONTRACT_PRAGMA_loop_body[] = "contracts:loop-id-body";
+const char CONTRACT_PRAGMA_loop_head[] = "contracts:loop-id-head";
+const char CONTRACT_PRAGMA_loop_latch[] = "contracts:loop-id-latch";
+const char CONTRACT_PRAGMA_loop_exiting[] = "contracts:loop-id-exiting";
+
```
These should remain private to the module

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.h:
```diff
@@ -34,6 +41,48 @@ class cfg_infot;
+// the outer loop would no longer observe the inner loop's write set,
+// but only corresponding `havoc` statements,
+// which can be instrumented in the usual way to achieve a "subset check".
+struct loop_graph_nodet : public graph_nodet<empty_edget>
```
can we move this to a separate `dfcc_instrument_loops.h` file/class?

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.h:
```diff
@@ -34,6 +41,48 @@ class cfg_infot;
+  loop_graph_nodet(
+    natural_loops_mutablet::loopt &loop,
+    const goto_programt::targett head,
+    const goto_programt::targett latch,
```
I've seen compute_natural_loops return loop that have more than one latch node, how do we handle this case ?

qinheping replied:
How about add a new latch node and let all original latch nodes jump to it instead.

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.h:
```diff
@@ -34,6 +41,48 @@ class cfg_infot;
+    natural_loops_mutablet::loopt &loop,
+    const goto_programt::targett head,
+    const goto_programt::targett latch,
+    const exprt &assigns,
```
isn't this supposed to be a vector/list  of `conditional_target_group_exprt`

remi-delmas-3000 replied:
Can we encapsulate the GOTO converter instance here ? Are these expressions supposed to be already GOTO converted or not ? When does goto_conversion happen for the loop contract expressions ?

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp:
```diff
@@ -361,6 +1121,87 @@ void dfcc_instrumentt::instrument_function_body(
+  local_may_aliast local_may_alias(goto_function);
+  // Iterate over the (natural) loops in the function, in topo-sorted order,
+  // and apply any loop contracts that we find.
+  for(const auto &idx : loop_nesting_graph.topsort())
```
Could we encapsulate this in a separate class, and then call that from here ?

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp:
```diff
@@ -361,6 +1121,87 @@ void dfcc_instrumentt::instrument_function_body(
+  local_may_aliast local_may_alias(goto_function);
+  // Iterate over the (natural) loops in the function, in topo-sorted order,
+  // and apply any loop contracts that we find.
+  for(const auto &idx : loop_nesting_graph.topsort())
```
Ideally our loop dependency graph should be acyclic, but we could have bugs that end up making it cyclic.
In case the graph is cyclic `topsort` will return an empty sequence. As sanity check we should check that the topsort has as many nodes as we added to the graph (it proves acyclicity).

qinheping replied:
Yes, we will add the check.

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp:
```diff
@@ -361,6 +1121,87 @@ void dfcc_instrumentt::instrument_function_body(
+      loop_node.head_target,
+      loop_node.latch_target,
+      cfg_info,
+      // don't skip any instructions
```
bad comment

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp:
```diff
@@ -361,6 +1121,87 @@ void dfcc_instrumentt::instrument_function_body(
+    size_t outer_idx = idx;
+    if(loop_node.out.size() != 0)
+    {
+      for(const auto &i : loop_nesting_graph.topsort())
```
is there a more direct way to access successors/predecessor nodes in the graph ? predecessor in topological order does not mean the nodes are actually connected

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp:
```diff
@@ -361,6 +1121,87 @@ void dfcc_instrumentt::instrument_function_body(
+      }
+    }
+
+    const exprt &outer_write_set =
+      (outer_idx == idx)
+        ? write_set
+        : loop_nesting_graph[outer_idx].write_set_symbol.symbol_expr();
```
I don't think this is necessarily pointing to an the enclosing loop (successor in topological order does not mean they are directly connected)

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp:
```diff
@@ -361,6 +1121,87 @@ void dfcc_instrumentt::instrument_function_body(
+  const auto &not_in_loop_filter =
+    [&](const goto_programt::instructiont::targett &target)
+  { return !is_loop_instruction(target); };
+
```
We also have to insert write set inclusion checks between the function's caller write_set and all top level loop write sets

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.h:
```diff
@@ -12,15 +12,22 @@ Author: Remi Delmas, delmasrd@amazon.com
 
 #include <goto-programs/goto_program.h>
 
+#include <analyses/local_may_alias.h>
+#include <analyses/natural_loops.h>
+#include <goto-instrument/loop_utils.h>
```
could we move the write set inference functions to a separate class e.g. `dfcc_instrument_loops.h`

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.h:
```diff
@@ -34,6 +41,48 @@ class cfg_infot;
+// We create a DAG representing nesting of various loops in goto_function,
+// sort them topologically, and instrument them in the top-sorted order.
+//
+// The goal here is not avoid explicit "subset checks" on nested write sets.
+// When instrumenting in a top-sorted order,
+// the outer loop would no longer observe the inner loop's write set,
+// but only corresponding `havoc` statements,
+// which can be instrumented in the usual way to achieve a "subset check".
```
this was how it used to work, but is not true in this implementation. Here we have explicit inner-outer loop and function-top-level-loop inclusion checks, and we tag all instructions added to instrument a given loop so they don't get instrumented again for the outer loop;

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.h:
```diff
@@ -34,6 +41,48 @@ class cfg_infot;
+  symbolt write_set_symbol;
+
+  loop_graph_nodet(
+    natural_loops_mutablet::loopt &loop,
```
Aren't these going to get invalidated by the instrumentation ? I would prefer to avoid storing these as trying to use them could lead to problems.
We should be using tags + search instead, i.e. we tag the loop instructions when constructing the dependency graph. To find the head, latch of loop number i  we search the instruction with these tags in the program.

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.h:
```diff
@@ -372,6 +422,51 @@ class dfcc_instrumentt
+    goto_programt &function_body);
+
+  /// Tags program instructions with pragmas
+  void tag_loop_instruction(
```
can we build the loop dep graph and tag instructions in single atomic operation ? Just to make sure there is no possibility to use outdated loop descriptors to do the tagging.

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.h:
```diff
@@ -372,6 +422,51 @@ class dfcc_instrumentt
+    assignst &assigns);
+
+  /// Add instructions for loop contracts.
+  void add_loop_contracts_instructions(
```
could we move this to a separate class and split it in `add_pre_header_instuctions` `add_step_instructions`, `add_body_instructions`, `add_exit_instructions` ?

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp:
```diff
@@ -338,6 +340,764 @@ void dfcc_instrumentt::instrument_function(
+    try
+    {
+      get_assigns_dfcc(local_may_alias, goto_function, idx, to_havoc);
+
```
@qinheping now that we are deprecating `__CPROVER_POINTER_OBJECT` in assigns clauses this function needs to widen targets using `__CPROVER_object_whole`

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp:
```diff
@@ -338,6 +371,137 @@ void dfcc_instrumentt::instrument_function(
+    const symbol_exprt &outer_write_set =
+      (outer_idx == idx) ? *write_set_ptr
+                         : loop_nesting_graph[outer_idx]
+                             .addr_of_loop_write_set_symbol.symbol_expr();
```
I believe you want to use `to_symbol_expr(write_set)` here as you may otherwise end up dereferencing a NULL pointer in `*write_set_ptr`.

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp:
```diff
@@ -338,6 +371,137 @@ void dfcc_instrumentt::instrument_function(
+    // Compute the write set of the outer loop. If there is no outer loop, use
+    // the write set of the function instead.
+    size_t outer_idx = idx;
+    if(loop_node.out.size() != 0)
```
Nit pick: use `!loop_node.out.empty()` instead of testing `size` against 0.

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_instrument.cpp:
```diff
@@ -0,0 +1,748 @@
+    // If we jump here, then the loop runs at least once,
+    // so add the base case assertion:
+    //   assert(initial_invariant_val)
+    code_assertt assertion{initial_invariant_val};
+    assertion.add_source_location() = loop_head_location;
+    assertion.add_source_location().set_comment(
+      "Check loop invariant before entry");
+    converter.goto_convert(assertion, step_instrs, symbol_mode);
```
What is our criterion for when we use `converter` vs when we directly add instructions to `step_instrs`? (I'd be much in favour of only using the latter.)

remi-delmas-3000 replied:
the contract expressions come directly from the C front end and we know they haven't been through goto conversion yet. So we have to apply conversion before inserting them in the program.

tautschnig replied:
Ok, I might have to take another look. It's just a bit confusing that sometimes we don't use the converter and other times we do. Kind of hoping that we are making the right decision each time.

remi-delmas-3000 replied:
the rule is to use the converter on contract expressions.

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_instrument.h:
```diff
@@ -0,0 +1,296 @@
+  dfcc_contract_clauses_codegent &contract_clauses_codegen;
+  namespacet ns;
+
+  /// @brief Return assigns extract from `assigns_clause`. Infer assigns targets
```
Here and everywhere else: we use \ instead of @ in our Doxygen comments.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_utils.cpp:
```diff
@@ -0,0 +1,514 @@
+  const irep_idt &instr_type,
+  const std::size_t loop_nesting_id)
+{
+  irept &irep = target->source_location_nonconst();
+  irep.set(ID_loop_instr_type, instr_type);
+  irep.set(ID_loop_nesting_id, loop_nesting_id);
+}
+
+irep_idt get_loop_instr_type(const goto_programt::instructiont::targett &target)
+{
+  irept &irep = target->source_location_nonconst();
+  return irep.get(ID_loop_instr_type);
+}
+
+std::size_t
+get_loop_nesting_id(const goto_programt::instructiont::targett &target)
+{
+  PRECONDITION(!get_loop_instr_type(target).empty());
+  irept &irep = target->source_location_nonconst();
```
Use `source_locationt` or `auto` instead of `irept`. (Yes, we inherit from that, but it's an implementation detail.)

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_utils.cpp:
```diff
@@ -0,0 +1,514 @@
+  if(!get_loop_instr_type(target).empty())
+  {
+    target->source_location_nonconst().remove(ID_loop_instr_type);
+    target->source_location_nonconst().remove(ID_loop_nesting_id);
```
Use `remove_loop_instr_type(target->source_location_nonconst());` since you have it...

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_utils.cpp:
```diff
@@ -0,0 +1,514 @@
+    target->source_location_nonconst().remove(ID_loop_instr_type);
+    target->source_location_nonconst().remove(ID_loop_nesting_id);
+
+    return;
```
Useless?

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_utils.cpp:
```diff
@@ -0,0 +1,514 @@
+        loop_end = t;
+    }
+
+    if(loop_end == loop_head)
+    {
+      log.error() << "Could not find end of the loop starting at: "
+                  << loop_head->source_location() << messaget::eom;
+      throw 0;
+    }
```
Should this be a `DATA_INVARIANT`? We throw exceptions when something might be actionable for the user, but I'm not sure what the user could do here?

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_utils.cpp:
```diff
@@ -0,0 +1,514 @@
+    // Check 2. no jumps into instructions in loops from outside.
+    for(const auto &i : loop_content)
+    {
+      if(std::distance(loop_head, i) < 0 || std::distance(i, loop_end) < 0)
```
How does this actually work? There is a lot of use of `auto` here (too much, IMHO) so I am guessing types: aren't `loop_head` and `i` std::list iterators? So the only way to compute the distance is increment-and-test. I don't see how a value of less-than-zero would be possible in that case?

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_utils.cpp:
```diff
@@ -0,0 +1,514 @@
+              i->output(mstream);
+              mstream << messaget::eom;
+            });
+          throw 0;
```
Avoid `throw 0`. Use `analysis_exceptiont`.

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_utils.cpp:
```diff
@@ -0,0 +1,514 @@
+    const auto idx = loop_nesting_graph.add_node(
+      loop_head, loop_end, assigns_clause, invariant, decreases_clause);
+
+    if(
+      assigns_clause.is_nil() && invariant.is_nil() &&
+      decreases_clause.is_nil())
+      continue;
```
Nit pick: braces around body after multi-line condition.

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_utils.cpp:
```diff
@@ -0,0 +1,514 @@
+        << " does not have contracts, but an enclosing loop does.\n"
+        << "Please provide contracts for this loop, or unwind it first."
+        << messaget::eom;
+      throw 0;
```
Use `analysis_exceptiont`.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_utils.cpp:
```diff
@@ -0,0 +1,514 @@
+        inner_loop_ids.begin(),
+        inner_loop_ids.end(),
+        get_loop_nesting_id(i_it)) == inner_loop_ids.end() ||
+      has_disable_assigns_check_pragma_dfcc(i_it))
+      continue;
```
Braces around body

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_utils.cpp:
```diff
@@ -0,0 +1,514 @@
+  }
+}
+
+void widen_assigns_dfcc(
+  assignst &assigns,
```
Use `assignst` as return type.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_utils.cpp:
```diff
@@ -0,0 +1,514 @@
+        args.emplace_back(address_of_expr);
+        result.emplace(side_effect_expr_function_callt(
+          widen_func_symbol,
+          args,
```
I think you could just use `{address_of_expr}` here in place.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1206,6 +1198,16 @@ void goto_instrument_parse_optionst::instrument_goto_program()
       cmdline.get_values("nondet-static-exclude").begin(),
       cmdline.get_values("nondet-static-exclude").end());
 
+    if(
+      cmdline.isset(FLAG_LOOP_CONTRACTS) &&
+      cmdline.isset(FLAG_LOOP_CONTRACTS_NO_UNWIND))
+    {
+      log.warning() << "**** WARNING: transformed loops will not be unwound "
+                    << "after applying loop contracts. Remember to unwind "
+                    << "them at least twice to pass unwinding-assertions."
+                    << messaget::eom;
+    }
```
Maybe this is a discussion we've had before and I just forgot about it: can you remind me as to why we don't just unwind those loops?

qinheping replied:
Unwinding transformed loops here will duplicate all assertions in the loops. And in some cases we don't want to duplicate assertions. For example, Kani assign every reachability check a unique ID, duplicate such assertion will result that two assertions have the same ID and fail Kani' CBMC-result rendering.
#7524

tautschnig replied:
Would you mind inserting a code comment documenting this thought process?

remi-delmas-3000 replied:
done

remi-delmas-3000 commented on src/goto-instrument/contracts/dynamic-frames/dfcc.cpp:
```diff
@@ -214,8 +212,22 @@ dfcct::dfcct(
+      library),
     max_assigns_clause_size(0)
 {
+  instrument.unwind_transformed_loops = unwind_transformed_loops;
```
we should not be passing configuration flags through side effects, this should be set once and for all at construction

remi-delmas-3000 replied:
fixed

qinheping commented:
Note that there are two problem not solved in this PR. We will create issues for them once this PR is merged.
1. `loop_assigns_inference-01` [[main.c](https://github.com/qinheping/cbmc/blob/loop-contracts-dfcc/regression/contracts-dfcc/loop_assigns_inference-01/main.c)]. As we don't inline function call in loops in dfcc, we will fail to infer targets assigned in callee functions. We can either inlining function calls, use local alias analysis, and then undo the inlining, or use global alias anlyais.
2. `loop_assigns-slice-assignable-scalar`[[main.c](https://github.com/qinheping/cbmc/blob/loop-contracts-dfcc/regression/contracts-dfcc/loop_assigns-slice-assignable-ptr/main.c)]. We have to use a stronger invariant saying that `b->data` stays unchanged. Otherwise, we will get error: `bv_to_expr` on `bool` is not supported.

Commit:
CONTRACTS: pointer predicates now use dfcc_cfg_infot

Commit:
CONTRACTS: avoid circular dependency between classes

Do not instrument code in `dfcc_spec_functionst` but
rather one layer above in `dfcc_contract_functionst`.
This allows to avoid creating a circular depedency on
`dfcc_instrumentt` later.

Commit:
CONTRACTS: fix missing source location

Commit:
CONTRACTS: add loop contracts support in DFCC

Commit:
CONTRACTS: update loop tests for DFCC

Commit:
CONTRACTS: loop normal form tests for DFCC

qinheping commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.h:
```diff
@@ -0,0 +1,53 @@
+  goto_programt::targett latch;
+
+  /// Set of loop instructions
+  const loopt instructions;
```
Let's try to explicitly use `const typename natural_loops_mutablet::loopt &instructions` to see if we can resolve the build error.

remi-delmas-3000 replied:
fixed

qinheping commented on src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.h:
```diff
@@ -0,0 +1,53 @@
+  dfcc_loop_nesting_graph_nodet(
+    goto_programt::targett head,
+    goto_programt::targett latch,
+    const loopt instructions)
```
```const typename natural_loops_mutablet::loopt &instructions```

remi-delmas-3000 replied:
fixed

Commit:
CONTRACTS: test combination of loop contracts and recursive functions

Commit:
CONTRACTS: new test with ternary assignment LHS

tautschnig reviewed (changes requested):
Please apply the script from the commit message in https://github.com/diffblue/cbmc/pull/7581/commits/01fd5ea12977c51333845cabf26a0d9acb865556 to de-duplicate tests.

Commit:
CONTRACTS: deduplicate tests

remi-delmas-3000 commented:
> Please apply the script from the commit message in [01fd5ea](https://github.com/diffblue/cbmc/commit/01fd5ea12977c51333845cabf26a0d9acb865556) to de-duplicate tests.

Deduplication done

tautschnig commented on regression/contracts-dfcc/invar_havoc_dynamic_multi-dim_array_all_const_idx/debug:
```diff
@@ -0,0 +1,2715 @@
+CBMC version 5.81.0 (cbmc-5.81.0-38-g8a964f611d-dirty) 64-bit x86_64 linux
```
What is this file about? Is this perhaps an unintentional add?

remi-delmas-3000 replied:
removed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc.h:
```diff
@@ -157,8 +163,7 @@ class dfcct
   /// \param to_check (function,contract) pair for contract checking
   /// \param allow_recursive_calls Allow the checked function to be recursive
   /// \param to_replace functions-to-contract mapping for replacement
-  /// \param apply_loop_contracts apply loop contract transformations iff true
-  /// havocing static program symbols.
+  /// \param loop_contract_mode
```
Wants some non-empty comment.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_contract_clauses_codegen.h:
```diff
@@ -8,8 +8,6 @@ Date: February 2023
 \*******************************************************************/
 
 /// \file
-/// Translates assigns and frees clauses of a function contract or
-/// loop contract into goto programs that build write sets or havoc write sets.
```
Is this removal intentional? If so, it should be replaced by something non-empty.

remi-delmas-3000 replied:
The `\file` tag is needed so that it is picked up by doxygen, but this text was moved to the class instead of the file.

I got tired of having file comments say "This file contains a class that does X, Y, Z" and have the class documentation say "this class does X, Y, Z".

remi-delmas-3000 replied:
the file and the class now both have the same comment again

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_contract_functions.cpp:
```diff
@@ -66,6 +69,29 @@ dfcc_contract_functionst::dfcc_contract_functionst(
   spec_functions.to_spec_frees_function(
     spec_frees_function_id, nof_frees_targets);
+
+  // instrument the generated functions for side-effect checking
+  std::set<irep_idt> function_pointer_contracts;
```
I don't quite understand how the comment fits with the declaration.

remi-delmas-3000 replied:
this comment applies to the two next lines of code.

remi-delmas-3000 replied:
deleted

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_contract_functions.cpp:
```diff
@@ -66,6 +69,29 @@ dfcc_contract_functionst::dfcc_contract_functionst(
+  // instrument the generated functions for side-effect checking
+  std::set<irep_idt> function_pointer_contracts;
+
+  instrument.instrument_function(
+    spec_assigns_function_id,
+    dfcc_loop_contract_modet::NONE,
+    function_pointer_contracts);
+
+  INVARIANT(
+    function_pointer_contracts.empty(),
+    id2string(spec_assigns_function_id) +
+      " shall not contain calls to " CPROVER_PREFIX "obeys_contract");
+
+  instrument.instrument_function(
+    spec_frees_function_id,
+    dfcc_loop_contract_modet::NONE,
+    function_pointer_contracts);
+
+  INVARIANT(
+    function_pointer_contracts.empty(),
+    id2string(spec_frees_function_id) +
+      " shall not contain calls to " CPROVER_PREFIX "obeys_contract");
```
I think we should have a variant of `instrument_function` that takes only two arguments and includes that `INVARIANT`.

remi-delmas-3000 replied:
factored

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_spec_functions.cpp:
```diff
@@ -272,13 +272,6 @@ void dfcc_spec_functionst::to_spec_assigns_function(
 
   goto_model.goto_functions.update();
 
-  // instrument for side-effects checking
-  std::set<irep_idt> function_pointer_contracts;
-  instrument.instrument_function(function_id, function_pointer_contracts);
-  INVARIANT(
-    function_pointer_contracts.empty(),
-    "discovered function pointer contracts unexpectedly");
-
```
We previously had that hot-fix re-instating this piece of code. Why is it now safe to remove this?

remi-delmas-3000 replied:
because the wrong removal was accidentally pulled from this commit

tautschnig reviewed (changes requested):
Seems ok to me except for the spurious `debug` file.

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_spec_functions.h:
```diff
@@ -32,28 +31,23 @@ class goto_modelt;
-/// dfcc library functions that actually built frame descriptions.
-/// The resulting function is then itself instrumented for frame condition
-/// checking to be able to prove the absence of side effects.
+/// into GOTO function that populate a write set instance or havoc a write set
```
```suggestion
/// into GOTO functions that populate a write set instance or havoc a write set
```

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_swap_and_wrap.cpp:
```diff
@@ -57,38 +56,45 @@ dfcc_swap_and_wrapt::dfcc_swap_and_wrapt(
   {
-    auto old_contract_id = pair.first->second.first;
-    auto old_contract_mode = pair.first->second.second;
+    irep_idt old_contract_id = pair.first->second.first;
+    dfcc_contract_modet old_contract_mode = pair.first->second.second.first;
+    dfcc_loop_contract_modet old_loop_contract_mode =
+      pair.first->second.second.second;
```
Could we create regression test to cover the lines below?

remi-delmas-3000 replied:
This is a sanity check that does not seem reachable given the checks that are already performed upstream and the fact that loop contracts are a global option, but it is still safe to keep it there.

feliperodri replied:
[Codecov](https://github.com/diffblue/cbmc/pull/7541/checks?check_run_id=13347318605) info says it's not covered. False alarm?

feliperodri commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1206,6 +1198,22 @@ void goto_instrument_parse_optionst::instrument_goto_program()
       cmdline.get_values("nondet-static-exclude").begin(),
       cmdline.get_values("nondet-static-exclude").end());
 
+    if(
+      cmdline.isset(FLAG_LOOP_CONTRACTS) &&
+      cmdline.isset(FLAG_LOOP_CONTRACTS_NO_UNWIND))
+    {
+      // When the model is produced by Kani, we must not automatically unwind
+      // the backjump introduced by the loop transformation.
+      // Automatic unwinding duplicates assertions found in the loop body, and
+      // since Kani expects property identifiers to remain unique. Having
+      // duplicate instances of the assertions makes Kani fail to handle the
+      // analysis results.
+      log.warning() << "**** WARNING: transformed loops will not be unwound "
+                    << "after applying loop contracts. Remember to unwind "
+                    << "them at least twice to pass unwinding-assertions."
+                    << messaget::eom;
+    }
```
Do we have a GitHub issue to track this?

remi-delmas-3000 replied:
It is not really an issue, it is just two modes of operation.

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp:
```diff
@@ -1249,3 +1238,49 @@ void dfcc_instrumentt::instrument_other(
+    const auto &loop = cfg_info.get_loop_info(loop_id);
+    if(loop.invariant.is_nil() && loop.decreases.empty())
+    {
+      // skip loops that do not have contracts
+      continue;
```
It'd be nice to have debug information about skipped loops. Not a blocker.

remi-delmas-3000 replied:
fixed

feliperodri reviewed (approved):
LGTM modulo Michael's comments.

Commit:
fix missing file doc
